## Упражнения по библиотеке Numpy

In [3]:
import numpy as np
import time


In [46]:
np.random.seed(1)
arr = np.random.randint(-10, 10, size=(10, 10))
arr

array([[ -5,   1,   2,  -2,  -1,   1,  -5,   5, -10,   6],
       [ -9,   2,  -3,   3,  -4,   8,  -5,   8,   1,   0],
       [  4,   8,  -6,  -1,   7, -10,   3,  -1,  -1,  -3],
       [ -9, -10,   7,  -2,   3,   9,   5,   0,  -2,  -3],
       [ -7,  -4,   7,  -7,  -6,   7,   1,   2,   6,   3],
       [  9,  -1,   8,   5, -10,  -6,   5,  -8,  -3,  -2],
       [ -1,  -7,  -3,  -6,  -5,   9,  -4,  -2, -10,  -8],
       [  0,   5,   5,  -3,   9,   0,   4, -10,  -9,   7],
       [  3,  -7, -10,   3,  -4,  -4,  -8,   2,   1,  -3],
       [  3,  -2,   1,   2,   1,  -6,  -3,  -3,   3,  -6]])

**1.** Дан случайный массив, поменять знак у элементов, значения которых между 3 и 8

In [47]:
arr[(arr > 3) & (arr < 8)] *= -1
arr

array([[ -5,   1,   2,  -2,  -1,   1,  -5,  -5, -10,  -6],
       [ -9,   2,  -3,   3,  -4,   8,  -5,   8,   1,   0],
       [ -4,   8,  -6,  -1,  -7, -10,   3,  -1,  -1,  -3],
       [ -9, -10,  -7,  -2,   3,   9,  -5,   0,  -2,  -3],
       [ -7,  -4,  -7,  -7,  -6,  -7,   1,   2,  -6,   3],
       [  9,  -1,   8,  -5, -10,  -6,  -5,  -8,  -3,  -2],
       [ -1,  -7,  -3,  -6,  -5,   9,  -4,  -2, -10,  -8],
       [  0,  -5,  -5,  -3,   9,   0,  -4, -10,  -9,  -7],
       [  3,  -7, -10,   3,  -4,  -4,  -8,   2,   1,  -3],
       [  3,  -2,   1,   2,   1,  -6,  -3,  -3,   3,  -6]])

**2.** Заменить максимальный элемент случайного массива на 0

In [48]:
indices = np.where(arr == np.max(arr))
arr[indices] = 0
arr

array([[ -5,   1,   2,  -2,  -1,   1,  -5,  -5, -10,  -6],
       [ -9,   2,  -3,   3,  -4,   8,  -5,   8,   1,   0],
       [ -4,   8,  -6,  -1,  -7, -10,   3,  -1,  -1,  -3],
       [ -9, -10,  -7,  -2,   3,   0,  -5,   0,  -2,  -3],
       [ -7,  -4,  -7,  -7,  -6,  -7,   1,   2,  -6,   3],
       [  0,  -1,   8,  -5, -10,  -6,  -5,  -8,  -3,  -2],
       [ -1,  -7,  -3,  -6,  -5,   0,  -4,  -2, -10,  -8],
       [  0,  -5,  -5,  -3,   0,   0,  -4, -10,  -9,  -7],
       [  3,  -7, -10,   3,  -4,  -4,  -8,   2,   1,  -3],
       [  3,  -2,   1,   2,   1,  -6,  -3,  -3,   3,  -6]])

**3.** Построить прямое произведение массивов (все комбинации с каждым элементом). На вход подается двумерный массив

In [190]:
arr = np.random.randint(-1, 5, size=(4,3))
print(arr)


[[ 3  2 -1]
 [ 2 -1  3]
 [ 2 -1  4]
 [ 2  0  3]]


In [189]:
np.array(np.meshgrid(arr.flatten(), arr.flatten())).T.reshape(-1, 2)

array([[2, 2],
       [2, 3],
       [2, 1],
       [2, 3],
       [2, 0],
       [2, 2],
       [3, 2],
       [3, 3],
       [3, 1],
       [3, 3],
       [3, 0],
       [3, 2],
       [1, 2],
       [1, 3],
       [1, 1],
       [1, 3],
       [1, 0],
       [1, 2],
       [3, 2],
       [3, 3],
       [3, 1],
       [3, 3],
       [3, 0],
       [3, 2],
       [0, 2],
       [0, 3],
       [0, 1],
       [0, 3],
       [0, 0],
       [0, 2],
       [2, 2],
       [2, 3],
       [2, 1],
       [2, 3],
       [2, 0],
       [2, 2]])

**4.** Даны 2 массива A (8x3) и B (2x2). Найти строки в A, которые содержат элементы из каждой строки в B, независимо от порядка элементов в B

In [220]:
A = np.array([[1, 2, 3],
              [4, 5, 6],
              [7, 8, 9],
              [3, 4, 2],
              [4, 7, 5],
              [6, 5, 4],
              [2, 9, 7],
              [5, 6, 2]])

B = np.array([[2, 4],
              [6, 5]])


In [224]:
A[np.where(np.all(np.isin(A,B),axis=1))]

array([[4, 5, 6],
       [6, 5, 4],
       [5, 6, 2]])

**5.** Дана 10x3 матрица, найти строки из неравных значений (например строка [2,2,3] остается, строка [3,3,3] удаляется)

In [246]:
A = np.array([[1, 2, 3],
                   [2, 2, 3],
                   [3, 3, 3],
                   [4, 4, 4],
                   [5, 5, 6],
                   [6, 6, 7],
                   [7, 7, 8],
                   [8, 8, 8],
                   [9, 9, 9],
                   [1, 2, 3]])
B = np.array([[1, 2, 3],
                   [1, 2,3]])

In [277]:
B = []
for i in A:
    row, count = np.unique(i,axis=0,return_counts=True)
    if 3 not in count:
        B.append(i)

np.vstack(B)

array([[1, 2, 3],
       [2, 2, 3],
       [5, 5, 6],
       [6, 6, 7],
       [7, 7, 8],
       [1, 2, 3]])

**6.** Дан двумерный массив. Удалить те строки, которые повторяются

In [278]:
A = np.array([[1, 2, 3],
                   [2, 2, 3],
                   [3, 3, 3],
                   [4, 4, 4],
                   [5, 5, 6],
                   [6, 6, 7],
                   [7, 7, 8],
                   [8, 8, 8],
                   [9, 9, 9],
                   [1, 2, 3]])

In [279]:
np.unique(A, axis = 0)

array([[1, 2, 3],
       [2, 2, 3],
       [3, 3, 3],
       [4, 4, 4],
       [5, 5, 6],
       [6, 6, 7],
       [7, 7, 8],
       [8, 8, 8],
       [9, 9, 9]])

______
______

Для каждой из следующих задач (1-5) нужно привести 2 реализации – одна без использования numpy (cчитайте, что там, где на входе или выходе должны быть numpy array, будут просто списки), а вторая полностью векторизованная с использованием numpy (без использования питоновских циклов/map/list comprehension).


__Замечание 1.__ Можно считать, что все указанные объекты непустые (к примеру, в __задаче 1__ на диагонали матрицы есть ненулевые элементы).

__Замечание 2.__ Для большинства задач решение занимает не больше 1-2 строк.

___

* __Задача 1__: Подсчитать произведение ненулевых элементов на диагонали прямоугольной матрицы.  
 Например, для X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]) ответ 3.

* Numpy

In [323]:
X = np.array([[4, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]])
Y = np.array([[4, 0, 1, 5], [2, 0, 2, 7], [3, 0, 3, 0]])

In [324]:
diagonal = np.diagonal(Y)
not_null_elem = diagonal[diagonal != 0]
np.prod(diagonal[diagonal != 0])

12

In [325]:
diagonal = np.diagonal(X)
not_null_elem = diagonal[diagonal != 0]
np.prod(diagonal[diagonal != 0])

12

* Python

In [326]:
X = [[4, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]
Y = [[4, 0, 1, 5], [2, 0, 2, 7], [3, 0, 3, 0]]

In [327]:
rng = min(len(Y), len(Y[0]))
prod_diag = 1
for i in range(rng):
    if Y[i][i] != 0:
        prod_diag *= Y[i][i]

print(prod_diag)

12


In [328]:
rng = min(len(X), len(X[0]))
prod_diag = 1
for i in range(rng):
    if X[i][i] != 0:
        prod_diag *= X[i][i]

print(prod_diag)

12


* __Задача 2__: Даны два вектора x и y. Проверить, задают ли они одно и то же мультимножество.  
  Например, для x = np.array([1, 2, 2, 4]), y = np.array([4, 2, 1, 2]) ответ True.

* Numpy

In [332]:
x = np.array([1, 2, 2, 4]) 
y = np.array([4, 2, 1, 2])

In [333]:
np.array_equal(np.sort(x), np.sort(y))

False

* Python

In [360]:
x = [1, 2, 2, 4]
y = [4, 2, 1, 2]

In [361]:
sorted(x) == sorted(y)

True

* __Задача 3__: Найти максимальный элемент в векторе x среди элементов, перед которыми стоит ноль. 
 Например, для x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0]) ответ 5. 

* Numpy

In [3]:
x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0])

In [4]:
indices = np.where(x == 0)[0]
indices = indices[indices < len(x)-1]
np.max(x[indices+1])

5

* Python


In [10]:
x = [6, 2, 0, 3, 0, 0, 5, 7, 0]

In [16]:
max_value = 0
for i in range(1, len(x)-1):
    if x[i-1] == 0 and x[i] > max_value:
        max_value = x[i]
print(max_value)

5


* __Задача 4__: Реализовать кодирование длин серий (Run-length encoding). Для некоторого вектора x необходимо вернуть кортеж из двух векторов одинаковой длины. Первый содержит числа, а второй - сколько раз их нужно повторить.  
 Например, для x = np.array([2, 2, 2, 3, 3, 3, 5]) ответ (np.array([2, 3, 5]), np.array([3, 3, 1])).

* Numpy

In [33]:
x = np.array([2, 2, 2, 3, 3, 3, 5, 2])

In [45]:
change_indices = np.where(np.diff(x) != 0)[0]+1
lengths = np.diff(np.concatenate(([0], change_indices, [len(x)])))
values = x[np.concatenate(([0], change_indices))]

(values, lengths)

(array([2, 3, 5, 2]), array([3, 3, 1, 1]))

* Python

In [48]:
x = [2, 2, 2, 3, 3, 3, 5,2]

In [49]:
i = 0
lengths = []
values = []
while (i <= len(x)-1):
        count = 1
        ch = x[i]
        values.append(x[i])
        j = i
        while (j < len(x)-1):
            if (x[j] == x[j+1]):
                count += 1
                j +=1 
            else:
                break
        lengths.append(count)
        i = j+1
(values, lengths)

([2, 3, 5, 2], [3, 3, 1, 1])

* __Задача 5__: Даны две выборки объектов - X и Y. Вычислить матрицу евклидовых расстояний между объектами. Сравните с функцией scipy.spatial.distance.cdist по скорости работы.

In [147]:
X = np.random.randint(10, size=(10, 3))
Y = np.random.randint(10, size=(8, 3))
Y

array([[6, 1, 0],
       [0, 4, 8],
       [5, 2, 2],
       [4, 1, 6],
       [9, 4, 5],
       [8, 3, 4],
       [6, 6, 9],
       [5, 4, 0]])

In [148]:
X

array([[7, 5, 9],
       [0, 5, 6],
       [0, 8, 9],
       [1, 2, 2],
       [7, 3, 8],
       [5, 7, 0],
       [8, 5, 6],
       [3, 0, 3],
       [1, 4, 7],
       [6, 0, 8]])

In [149]:
from scipy.spatial.distance import cdist 
start = time.time()
print(cdist(X, Y,metric='euclidean'))
end = time.time()
print()
print (end - start)

[[ 9.89949494  7.14142843  7.87400787  5.83095189  4.58257569  5.47722558
   1.41421356  9.2736185 ]
 [ 9.38083152  2.23606798  7.07106781  5.65685425  9.11043358  8.48528137
   6.78232998  7.87400787]
 [12.88409873  4.12310563 10.48808848  8.60232527 10.63014581 10.67707825
   6.32455532 11.04536102]
 [ 5.47722558  6.40312424  4.          5.09901951  8.77496439  7.34846923
   9.48683298  4.89897949]
 [ 8.30662386  7.07106781  6.40312424  4.12310563  3.74165739  4.12310563
   3.31662479  8.30662386]
 [ 6.08276253  9.89949494  5.38516481  8.54400375  7.07106781  6.40312424
   9.11043358  3.        ]
 [ 7.48331477  8.30662386  5.83095189  5.65685425  1.73205081  2.82842712
   3.74165739  6.78232998]
 [ 4.35889894  7.07106781  3.          3.31662479  7.48331477  5.91607978
   9.          5.38516481]
 [ 9.11043358  1.41421356  6.70820393  4.35889894  8.24621125  7.68114575
   5.74456265  8.06225775]
 [ 8.06225775  7.21110255  6.40312424  3.          5.83095189  5.38516481
   6.08276253  9.

* Numpy

In [150]:
start = time.time()
distances = np.sqrt(np.sum((X[:, np.newaxis] - Y) ** 2, axis=2))
print(distances)
end = time.time()
print()
print(end - start)

[[ 9.89949494  7.14142843  7.87400787  5.83095189  4.58257569  5.47722558
   1.41421356  9.2736185 ]
 [ 9.38083152  2.23606798  7.07106781  5.65685425  9.11043358  8.48528137
   6.78232998  7.87400787]
 [12.88409873  4.12310563 10.48808848  8.60232527 10.63014581 10.67707825
   6.32455532 11.04536102]
 [ 5.47722558  6.40312424  4.          5.09901951  8.77496439  7.34846923
   9.48683298  4.89897949]
 [ 8.30662386  7.07106781  6.40312424  4.12310563  3.74165739  4.12310563
   3.31662479  8.30662386]
 [ 6.08276253  9.89949494  5.38516481  8.54400375  7.07106781  6.40312424
   9.11043358  3.        ]
 [ 7.48331477  8.30662386  5.83095189  5.65685425  1.73205081  2.82842712
   3.74165739  6.78232998]
 [ 4.35889894  7.07106781  3.          3.31662479  7.48331477  5.91607978
   9.          5.38516481]
 [ 9.11043358  1.41421356  6.70820393  4.35889894  8.24621125  7.68114575
   5.74456265  8.06225775]
 [ 8.06225775  7.21110255  6.40312424  3.          5.83095189  5.38516481
   6.08276253  9.

* Python

In [132]:
X = [[7, 5, 9],
       [0, 5, 6],
       [0, 8, 9],
       [1, 2, 2],
       [7, 3, 8],
       [5, 7, 0],
       [8, 5, 6],
       [3, 0, 3],
       [1, 4, 7],
       [6, 0, 8]]

Y  = [[6, 1, 0],
       [0, 4, 8],
       [5, 2, 2],
       [4, 1, 6],
       [9, 4, 5],
       [8, 3, 4],
       [6, 6, 9],
       [5, 4, 0]]

In [151]:
def euclidean_distance(x, y):
    squared_diff = sum((i - j) ** 2 for i, j in zip(x, y))
    return squared_diff ** 0.5

start = time.time()
distances = [[euclidean_distance(x, y) for y in Y] for x in X]
end = time.time()

for i in distances:
    print(i)
    
print()
print(end - start)

[9.899494936611665, 7.14142842854285, 7.874007874011811, 5.830951894845301, 4.58257569495584, 5.477225575051661, 1.4142135623730951, 9.273618495495704]
[9.38083151964686, 2.23606797749979, 7.0710678118654755, 5.656854249492381, 9.1104335791443, 8.48528137423857, 6.782329983125268, 7.874007874011811]
[12.884098726725126, 4.123105625617661, 10.488088481701515, 8.602325267042627, 10.63014581273465, 10.677078252031311, 6.324555320336759, 11.045361017187261]
[5.477225575051661, 6.4031242374328485, 4.0, 5.0990195135927845, 8.774964387392123, 7.3484692283495345, 9.486832980505138, 4.898979485566356]
[8.306623862918075, 7.0710678118654755, 6.4031242374328485, 4.123105625617661, 3.7416573867739413, 4.123105625617661, 3.3166247903554, 8.306623862918075]
[6.082762530298219, 9.899494936611665, 5.385164807134504, 8.54400374531753, 7.0710678118654755, 6.4031242374328485, 9.1104335791443, 3.0]
[7.483314773547883, 8.306623862918075, 5.830951894845301, 5.656854249492381, 1.7320508075688772, 2.828427124

_______
________

* #### __Задача 6__: CrunchieMunchies __*__

Вы работаете в отделе маркетинга пищевой компании MyCrunch, которая разрабатывает новый вид вкусных, полезных злаков под названием **CrunchieMunchies**.

Вы хотите продемонстрировать потребителям, насколько полезны ваши хлопья по сравнению с другими ведущими брендами, поэтому вы собрали данные о питании нескольких разных конкурентов.

Ваша задача - использовать вычисления Numpy для анализа этих данных и доказать, что ваши **СrunchieMunchies** - самый здоровый выбор для потребителей.


In [77]:
import numpy as np

1. Просмотрите файл cereal.csv. Этот файл содержит количества калорий для различных марок хлопьев. Загрузите данные из файла и сохраните их как calorie_stats.

In [6]:
calorie_stats = np.loadtxt("./data/cereal.csv", delimiter=",")
calorie_stats

array([ 70., 120.,  70.,  50., 110., 110., 110., 130.,  90.,  90., 120.,
       110., 120., 110., 110., 110., 100., 110., 110., 110., 100., 110.,
       100., 100., 110., 110., 100., 120., 120., 110., 100., 110., 100.,
       110., 120., 120., 110., 110., 110., 140., 110., 100., 110., 100.,
       150., 150., 160., 100., 120., 140.,  90., 130., 120., 100.,  50.,
        50., 100., 100., 120., 100.,  90., 110., 110.,  80.,  90.,  90.,
       110., 110.,  90., 110., 140., 100., 110., 110., 100., 100., 110.])

2. В одной порции CrunchieMunchies содержится 60 калорий. Насколько выше среднее количество калорий у ваших конкурентов?

Сохраните ответ в переменной average_calories и распечатайте переменную в терминале

In [156]:
average_calories = np.mean(calorie_stats)
average_calories

106.88311688311688

3. Корректно ли среднее количество калорий отражает распределение набора данных? Давайте отсортируем данные и посмотрим.

Отсортируйте данные и сохраните результат в переменной calorie_stats_sorted. Распечатайте отсортированную информацию

In [157]:
calorie_stats_sorted = np.sort(calorie_stats)
calorie_stats_sorted

array([ 50.,  50.,  50.,  70.,  70.,  80.,  90.,  90.,  90.,  90.,  90.,
        90.,  90., 100., 100., 100., 100., 100., 100., 100., 100., 100.,
       100., 100., 100., 100., 100., 100., 100., 100., 110., 110., 110.,
       110., 110., 110., 110., 110., 110., 110., 110., 110., 110., 110.,
       110., 110., 110., 110., 110., 110., 110., 110., 110., 110., 110.,
       110., 110., 110., 110., 120., 120., 120., 120., 120., 120., 120.,
       120., 120., 120., 130., 130., 140., 140., 140., 150., 150., 160.])

4. Похоже, что большинство значений выше среднего. Давайте посмотрим, является ли медиана наиболее корректным показателем набора данных.

Вычислите медиану набора данных и сохраните свой ответ в median_calories. Выведите медиану, чтобы вы могли видеть, как она сравнивается со средним значением.

In [158]:
median_calories = np.median(calorie_stats)
median_calories

110.0

Среденее - сумма всех значений на количество. Медиана - занчение, кторое делит набор данных на две половины. При нечетном количестве данных: значение посередине, при четном: среднее значение двух значений, находящихся посередине.
Среднее - лучше отражает все значения выборки,так как при расчете использует все значение ее. Но не устойчива к выбросам. Медиана же устойчива к выбросам. 

5. В то время как медиана показывает, что по крайней мере половина наших значений составляет более 100 калорий, было бы более впечатляюще показать, что значительная часть конкурентов имеет более высокое количество калорий, чем CrunchieMunchies.

Рассчитайте различные процентили и распечатайте их, пока не найдете наименьший процентиль, превышающий 60 калорий. Сохраните это значение в переменной nth_percentile.

In [14]:
first_percentile=np.percentile(calorie_stats,30)
second_percentile=np.percentile(calorie_stats,75)
third_percentile=np.percentile(calorie_stats,4)
nth_percentile = 4
print(first_percentile)
print(second_percentile)
print(third_percentile)

100.0
110.0
70.0


6. Хотя процентиль показывает нам, что у большинства конкурентов количество калорий намного выше, это неудобная концепция для использования в маркетинговых материалах.

Вместо этого давайте подсчитаем процент хлопьев, в которых содержится более 60 калорий на порцию. Сохраните свой ответ в переменной more_calories и распечатайте его

In [15]:
more_calories = 100 - nth_percentile
more_calories

96

7. Это действительно высокий процент. Это будет очень полезно, когда мы будем продвигать CrunchieMunchies. Но один вопрос заключается в том, насколько велики различия в наборе данных? Можем ли мы сделать обобщение, что в большинстве злаков содержится около 100 калорий или разброс еще больше?

Рассчитайте величину отклонения, найдя стандартное отклонение, Сохраните свой ответ в calorie_std и распечатайте на терминале. Как мы можем включить эту ценность в наш анализ?

In [164]:
calorie_std=np.std(calorie_stats)
calorie_std

19.35718533390827

Нормальное отклонение показывает на сколько данные в наборе данных разбросаны относительно среднего значения. Если значение мало, то сконцентрированы вокруг среднего, если велико, то далеко от среднего. 
Здесь же отклонение относительно малое, что показывает, что данные в основном сконцентрированы близко к среднему.

8. Напишите короткий абзац, в котором кратко изложите свои выводы и то, как, по вашему мнению, эти данные могут быть использованы в интересах Mycrunch при маркетинге CrunchieMunchies.

Используя такие метрики как среднее, процентиль и нормальное отклонение, можно показать, что калорийность CrunchieMunchies меньше чем у конкурентов. Так как среднее = 107 показывает, что калорийность хлопьев у большинства конкурентов больше чем 60 калорий. Стандартное же отклонение, подвердило гипотезу, что калорийность у конкурентов больше чем 60 и близко к среднему. 